src/Shared/FileUtilities.cs

```diff
@@ -930,5 +946,21 @@ private static string CurrentExecutableOverride
                 return s_currentExecutableOverride;
             }
         }
+
+        /// <summary>
+        /// Returns true when the entry point application is Visual Studio.
+        /// </summary>
+        internal static bool RunningInVisualStudio
+        {
+            get
+            {
+                // Check if initialized and do so if not yet
+                if (!s_runningInVisualStudio.HasValue)
+                {
+                    GetTestExecutionInfo();
```

Comment:
I think the class could definitely use some improvements, but I don't think there are any issues with thread safety. There's very little state and the methods that set state look to be idempotent.